src/passes/Heap2Local.cpp

```diff
@@ -0,0 +1,611 @@
+
+    Rewriter(StructNew* allocation, Function* func, Module* module)
+      : allocation(allocation), func(func), builder(*module),
+        fields(allocation->type.getHeapType().getStruct().fields) {}
+
+    // We must track all the local.sets that write the allocation, to verify
+    // exclusivity.
+    std::unordered_set<LocalSet*> sets;
+
+    // All the expressions we reached during the flow analysis. That is exactly
+    // all the places where our allocation is used. We track these so that we
+    // can fix them up at the end, if the optimization ends up possible.
+    std::unordered_set<Expression*> reached;
+
+    // Maps indexes in the struct to the local index that will replace them.
+    std::vector<Index> localIndexes;
+
+    void applyOptimization() {
+      // Allocate locals to store the allocation's fields in.
+      for (auto field : fields) {
+        localIndexes.push_back(builder.addVar(func, field.type));
+      }
+
+      // Replace the things we need to using the visit* methods.
+      walk(func->body);
+    }
+
+    void visitLocalSet(LocalSet* curr) {
+      if (!reached.count(curr)) {
+        return;
+      }
+
+      // We don't need any sets of the reference to any of the locals it
+      // originally was written to.
+      if (curr->isTee()) {
+        replaceCurrent(curr->value);
+      } else {
+        replaceCurrent(builder.makeDrop(curr->value));
+      }
+    }
+
+    void visitStructNew(StructNew* curr) {
+      if (curr != allocation) {
+        return;
+      }
+
+      // We do not remove the allocation itself here, rather we make it
+      // unnecessary, and then depend on other optimizations to clean up. (We
+      // cannot simply remove it because we need to replace it with something of
+      // the same non-nullable type.)
+
+      // First, assign the initial values to the new locals.
+      std::vector<Expression*> contents;
+
+      if (!allocation->isWithDefault()) {
+        // We must assign the initial values to temp indexes, then copy them
+        // over all at once. If instead we did set them as we go, then we might
+        // hit a problem like this:
+        //
+        //  (local.set X (new_X))
+        //  (local.set Y (block (result ..)
+        //                 (.. (local.get X) ..) ;; returns new_X, wrongly
+        //                 (new_Y)
+        //               )
+        //
+        // Note how we assign to the local X and use it during the assignment to
+        // the local Y - but we should still see the old value of X, not new_X.
+        // Temp locals X', Y' can ensure that:
+        //
+        //  (local.set X' (new_X))
+        //  (local.set Y' (block (result ..)
+        //                  (.. (local.get X) ..) ;; returns the proper, old X
+        //                  (new_Y)
+        //                )
+        //  ..
+        //  (local.set X (local.get X'))
+        //  (local.set Y (local.get Y'))
+        std::vector<Index> tempIndexes;
```

Comment:
I guess we don't need to do this temp local thing when there's only one field.. But I guess other local opt passes will handle it anyway?